tests/unit/test_upgrade.py

```diff
@@ -0,0 +1,240 @@
+
+    # removes get_leader patch
+    mocker.stopall()
+
+    with pytest.raises(ClusterNotReadyError):
+        harness.charm.upgrade.pre_upgrade_check()
+
+
+def test_pre_upgrade_check_succeeds(harness, mocker):
+    mocker.patch.object(
+        ZooKeeperManager, "members_broadcasting", new_callable=PropertyMock, return_value=True
+    )
+    mocker.patch.object(
+        ZooKeeperManager, "server_members", new_callable=PropertyMock, return_value=[0]
+    )
+    mocker.patch.object(
+        ZooKeeperManager, "members_syncing", new_callable=PropertyMock, return_value=False
+    )
+    mocker.patch.object(ZooKeeperCluster, "stable", new_callable=PropertyMock, return_value=True)
+
+    harness.charm.upgrade.pre_upgrade_check()
+
+
+def test_build_upgrade_stack(harness):
+    with harness.hooks_disabled():
+        harness.add_relation_unit(harness.charm.peer_relation.id, f"{CHARM_KEY}/1")
+        harness.update_relation_data(
+            harness.charm.peer_relation.id, f"{CHARM_KEY}/1", {"private-address": "111.111.111"}
+        )
+        harness.add_relation_unit(harness.charm.peer_relation.id, f"{CHARM_KEY}/2")
+        harness.update_relation_data(
+            harness.charm.peer_relation.id, f"{CHARM_KEY}/2", {"private-address": "222.222.222"}
+        )
+        harness.add_relation_unit(harness.charm.peer_relation.id, f"{CHARM_KEY}/3")
+        harness.update_relation_data(
+            harness.charm.peer_relation.id, f"{CHARM_KEY}/3", {"private-address": "333.333.333"}
+        )
+
+    stack = harness.charm.upgrade.build_upgrade_stack()
+
+    assert stack[0] == 0
+    assert len(stack) == 4
+
+
+def test_zookeeper_dependency_model():
+    assert sorted(ZooKeeperDependencyModel.__fields__.keys()) == sorted(DEPENDENCIES.keys())
+
+    for value in DEPENDENCIES.values():
+        assert DependencyModel(**value)
+
+
+def test_upgrade_granted_sets_failed_if_failed_snap(harness, mocker):
+    mocker.patch.object(ZooKeeperSnap, "stop_snap_service")
+    mocker.patch.object(ZooKeeperSnap, "restart_snap_service")
+    mocker.patch.object(ZooKeeperSnap, "install", return_value=False)
+    mocker.patch.object(ZooKeeperUpgrade, "pre_upgrade_check")
+    mocker.patch.object(ZooKeeperUpgrade, "set_unit_completed")
+    mocker.patch.object(ZooKeeperUpgrade, "set_unit_failed")
+
+    mock_event = mocker.MagicMock()
+
+    harness.charm.upgrade._on_upgrade_granted(mock_event)
+
+    ZooKeeperSnap.stop_snap_service.assert_called_once()
+    ZooKeeperSnap.install.assert_called_once()
+    ZooKeeperSnap.restart_snap_service.assert_not_called()
+    ZooKeeperUpgrade.set_unit_completed.assert_not_called()
+    ZooKeeperUpgrade.set_unit_failed.assert_called_once()
+
+
+def test_upgrade_granted_sets_failed_if_failed_upgrade_check(harness, mocker):
+    mocker.patch.object(ZooKeeperSnap, "stop_snap_service")
+    mocker.patch.object(ZooKeeperSnap, "restart_snap_service")
+    mocker.patch.object(ZooKeeperSnap, "install", return_value=False)
```

Comment:
I think the `return_value` should be `True` here.

Reply:
What a good catch! Have updated :)